src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/Dependencies/Subscriptions/DependencyRulesSubscriber.cs

```diff
@@ -229,23 +216,18 @@ private IReadOnlyCollection<string> GetWatchedRules(RuleHandlerType handlerType)
                 return;
             }
 
+            if (!projectUpdate.ProjectChanges.Any(x => x.Value.Difference.AnyChanges))
+            {
+                return;
+            }
+
             // Create an object to track dependency changes.
             var changesBuilder = new CrossTargetDependenciesChangesBuilder();
 
             // Give each handler a chance to register dependency changes.
             foreach (Lazy<IDependenciesRuleHandler, IOrderPrecedenceMetadataView> handler in _handlers)
             {
-                ImmutableHashSet<string> handlerRules = handler.Value.GetRuleNames(handlerType);
-
-                // Slice project changes to include only rules the handler claims an interest in.
-                var projectChanges = projectUpdate.ProjectChanges
-                    .Where(x => handlerRules.Contains(x.Key))
-                    .ToImmutableDictionary();
-
-                if (projectChanges.Any(x => x.Value.Difference.AnyChanges))
-                {
-                    handler.Value.Handle(projectChanges, targetFrameworkToUpdate, changesBuilder);
-                }
+                handler.Value.Handle(projectUpdate.ProjectChanges, targetFrameworkToUpdate, changesBuilder);
```

Comment:
Do handles expect this change behavior (previously only got changes related to them, now get everything)?

Reply:
Yes, handlers look up via their rules directly and ignore extra members, so it's equivalent in that regard. The behaviour change here is that _all_ handlers are now called regardless of whether they have any changes to process. The handlers themselves deal with this. Currently they allocate enumerators for added/changed/removed, but in a future change I'll protect against that by checking for zero counts before each `foreach`.

Currently there are two handler implementations. One in `DependenciesRuleHandlerBase` and an override of that which is specific to packages in `PackageRuleHandler`. They are very similar methods and I'm currently looking to merge them.

It's this handler where the logic we discussed on Monday will most likely live, hence my wanting to get all the bits leading up to this nice and clean.